### PR TITLE
make groups cv splitters and merge to Galaxy-ML

### PR DESCRIPTION
**Pull request by qiagu**
1) training dataset load as float
~~2) change backend and UI to support multiple-label classification~~
3) add column selector to groups cv splitters, like GroupKFold, LeaveOneGroupOut _etc._
4) rewrite the model validation tool

### CODE REVIEW

File: tools/sklearn/feature_selection.xml [hunk 1]
```diff
@@ -50,6 +50,13 @@ params['fs_algorithm_selector']['estimator_selector']['c_estimator'] =\
         '$fs_algorithm_selector.estimator_selector.c_estimator'
 #end if
 
+#if $fs_algorithm_selector.selected_algorithm == 'RFECV'\
+        and $fs_algorithm_selector.options.cv_selector.selected_cv\
+        in ['GroupKFold', 'GroupShuffleSplit', 'LeaveOneGroupOut', 'LeavePGroupsOut']:
+params['fs_algorithm_selector']['options']['cv_selector']['groups_selector']['infile_g'] =\
```
[the remark this answers is not in the document]
Comment (qiagu) [reply tools/sklearn/feature_selection.xml]: ?? Could you please explain a little more?

File: tools/sklearn/main_macros.xml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 <macros>
-  <token name="@VERSION@">1.0</token>
+  <token name="@VERSION@">1.0.1</token>
```
[the remark this answers is not in the document]
Comment (qiagu) [reply tools/sklearn/main_macros.xml]: I don't understand the benefit of `1.0.0.1`, but I believe you have good reasons.

File: tools/sklearn/model_validation.xml [hunk 3]
```diff
@@ -64,133 +75,100 @@ y = read_columns(
...
-
-## Set up feature selector and add to pipeline steps.
-if params['feature_selection']['do_feature_selection'] == 'Yes':
-    feature_selector = feature_selector(params['feature_selection']['fs_algorithm_selector'])
-    pipeline_steps.append( ('feature_selector', feature_selector) )
-
-## Set up estimator and add to pipeline.
-estimator_json = params['model_validation_functions']['estimator_selector']
-estimator = get_estimator(estimator_json)
+# load pipeline
```
[the remark this answers is not in the document]
Comment (qiagu) [reply tools/sklearn/model_validation.xml]: I know `##` is recommended comment format for Cheetah. However, in my experience, `#` doesn't harm anything, although it may be tried to be executed.

File: tools/sklearn/model_validation.xml [hunk 4]
```diff
@@ -200,64 +178,34 @@ rval.to_csv(path_or_buf='$outfile', sep='\t', header=False, index=False)
...
+                                <add value="["/>
+                                <add value="]"/>
+                            </valid>
+                        </sanitizer>
+                    </param>
+                    <param argument="exploit_incremental_learning" type="boolean" optional="true" truevalue="booltrue" falsevalue="boolfalse" checked="false" help="Whether to apply incremental learning to speed up fitting of the estimator if supported"/>
                     <expand macro="pre_dispatch"/>
                     <expand macro="shuffle" checked="false" label="shuffle" help="Whether to shuffle training data before taking prefixes"/>
-                    <expand macro="random_state"/>
+                    <expand macro="random_state" help="If int, the seed used by the random number generator. Used when `shuffle` is True"/>
```
[the remark this answers is not in the document]
Comment (qiagu) [reply tools/sklearn/model_validation.xml]: Fixed, thanks.

File: tools/sklearn/search_model_validation.xml [hunk 5]
```diff
@@ -11,8 +11,13 @@
     <version_command>echo "@VERSION@"</version_command>
     <command>
         <![CDATA[
+        #if $search_schemes.options.cv_selector.selected_cv in ['GroupKFold', 'GroupShuffleSplit', 'LeaveOneGroupOut', 'LeavePGroupsOut']:
+        inputs="$inputs,$search_schemes.options.cv_selector.groups_selector.infile_g";
```
[the remark this answers is not in the document]
Comment (qiagu) [reply tools/sklearn/search_model_validation.xml]: You're right! Do you want me to change it?

[the remark this answers is not in the document]
Comment (bgruening) [reply tools/sklearn/search_model_validation.xml]: That would be nice.

**Comment by qiagu** [comment]
Once conda environment changes, many compare-size tests for picked objects fail. :(

**Comment by bgruening** [comment]
Is that because of imbalanced-learn?
You can run `planemo test --update_test_data` but be aware this will update all your test data :)

**Comment by bgruening** [comment]
Uha, I will review today and try to get the conda install working.

ping @anuprulez

**Comment by qiagu** [comment]
It seems the recent change makes the searchcv tool much slower. Two possible causes, 1) more copies of conda activation blocks in the `tool_script.sh` due to adding dependencies, 2) the extra layer of `try_get_attr` fucntion. I'm still working on the profiling though.

**Comment by bgruening** [comment]
How do you measure this? It's not likely to be conda, conda is a one time setup.

File: tools/sklearn/main_macros.xml [hunk 6]
```diff
@@ -1,14 +1,17 @@
...
       <requirements>
           <requirement type="package" version="3.6">python</requirement>
-          <requirement type="package" version="0.20.2">scikit-learn</requirement>
-          <requirement type="package" version="0.23.4">pandas</requirement>
+          <requirement type="package" version="0.20.3">scikit-learn</requirement>
+          <requirement type="package" version="0.24.2">pandas</requirement>
           <requirement type="package" version="0.80">xgboost</requirement>
           <requirement type="package" version="0.9.13">asteval</requirement>
-          <yield />
+          <requirement type="package" version="0.6">skrebate</requirement>
```
[the remark this answers is not in the document]
Comment (qiagu) [reply tools/sklearn/main_macros.xml]: I am fine to go back to the old scheme. The feature_selection, model_validation, pipeline and searchcv tools may need one unified library.

File: tools/sklearn/model_validation.xml [hunk 7]
```diff
@@ -15,15 +15,27 @@
...
+    cluster, compose, decomposition, ensemble, feature_extraction,
+    feature_selection, gaussian_process, kernel_approximation, metrics,
+    model_selection, naive_bayes, neighbors, pipeline, preprocessing,
+    svm, linear_model, tree, discriminant_analysis)
 
-exec(open('$__tool_directory__/utils.py').read(), globals())
+sys.path.insert(0, '$__tool_directory__')
+from utils import SafeEval, get_cv, get_scoring, load_model, read_columns
+
+N_JOBS = int(__import__('os').environ.get('GALAXY_SLOTS', 1))
```
[the remark this answers is not in the document]
Comment (qiagu) [reply tools/sklearn/model_validation.xml]: I thought having os in the namespace was not safe, but may not be the case.

**Comment by qiagu** [comment]
> How do you measure this? It's not likely to be conda, conda is a one time setup.

https://github.com/galaxyproject/galaxy/issues/7849
